improve new gene panel documentation and parsing, fix missing inheritance models bug (#2436)
Clinical-Genomics/scout: docs/user-guide/panels.md

### File format

> The gene panel is semicolon (;) or tab-separated text file with an optional header that describes the columns and one line for each gene entry.

Reviewer: ```suggestion
The gene panel is a tab, space or semicolon (;)-separated text file with an optional header that describes the columns and one line for each gene entry. You must use the same delimiter for the whole file. Ideally do not use the delimiter characters in other places in the file. Consult with an admin if you need to use the delimiter characters in other fields for help with escaping them or ensuring a higher priority separator is used on a previous line.
```

Author: Are you sure that space is allowed? I don't think so. I think I've tried one with space from a customer some time ago and it  didn't work

Reviewer: So it should, but it is a little tricky in that if there is a tab somewhere, that will take precedence. More eyerollingly, both have priority over semicolon. See https://github.com/Clinical-Genomics/scout/blob/41b61ba112ed685e39a5029b2d7491643fbe1081/scout/parse/panel.py#L174 and following.

Reviewer: If documentation starts getting ridiculous, it is a sign that the corresponding legacy code needs a refactor.

Author: > If documentation starts getting ridiculous, it is a sign that the corresponding legacy code needs a refactor.

Perhaps something that checks the file line by line and tells you what is wrong with that line?

Author: What about removing the space as a separator directly in the code?

Reviewer: That would very likely solve more problems than it would create. If so, check test, fixture and legacy files to be sure.

Reviewer: Yeah or just remake/use a regular csv/tsv parser class already that properly handles escapes, delimiters, quotes etc.

> > If documentation starts getting ridiculous, it is a sign that the corresponding legacy code needs a refactor.
> 
> 
> 
> Perhaps something that checks the file line by line and tells you what is wrong with that line?

Reviewer: (No offense previous authors - I've done worse hacks, and I'm also a little hungry now.. 😅)

Author: Ok, let's revise the parsing, tomorrow!

Author: The only place where space as a separator was mentioned and used is in the panel parser, and even there it was not contemplated in the comments, see [L202](https://github.com/Clinical-Genomics/scout/blob/41b61ba112ed685e39a5029b2d7491643fbe1081/scout/parse/panel.py#L202).

I've removed it from the parsing now and it should be fine I think!

Author: @dnil I'm going to merge this is you are OK with it!

Reviewer: I think this is good, but perhaps ping one or two panel maintainers at CMMS to be sure this doesn't disrupt their workflow? 😊

Author: Asked for feedback: CMMS downloads the demo excel file, pasts genes into it and the saves it as a csv, I'll try the procedure with this branch on stage and it it works I'll merge!

Author: And of course there are issues, because if you take and excel and you save it as a csv file then the separator is comma! Tried to save it as a tab separated file and the result is better, it just adds annoying quotes around the fields that have comma:

![image](https://user-images.githubusercontent.com/28093618/111302603-e6f2a000-8653-11eb-9746-96ff8b6f3c1c.png)

OMG I hate Excel, I'll fix it so it removes these quotes during the parsing

Author: Tested the latest changes:
- [ ] Download example excel file
- [ ] edit
- [ ] Save it as tab separated
- [ ] Import it as a new panel

Works and the quotes are gone: 

![image](https://user-images.githubusercontent.com/28093618/111306014-f70c7e80-8657-11eb-832a-a3cabc679b2b.png)

Reviewer: > And of course there are issues, because if you take an excel and you save it as a csv file then the separator is comma! Tried to save it as a tab separated file and the result is better, it just adds annoying quotes around the fields that have comma:
> 
> ![image](https://user-images.githubusercontent.com/28093618/111302603-e6f2a000-8653-11eb-9746-96ff8b6f3c1c.png)
> 
> OMG I hate Excel, I'll fix it so it removes these quotes during the parsing

Agree on that. And comma or semicolon depends on locale. Its not very convenient with comma separation if you have one in every decimal number.. 😆

Reviewer: Again, there are loads of csv/tsv parsing libraries. One more dependency, but.. 🤷

Author: I suggest let's keep it this way and then if there are still problems we change how we do the parsing. I don't have the energy to completely re-write it now.

> The columns that will be used by scout are the following. **Please note that the order of the columns can't be changed**

Reviewer: ```suggestion
The columns that will be used by scout are the following. **Please note that if you do not include a header, the order of the columns can't be changed**
```

> The columns that will be used by scout are the following. **Please note that if you do not include a header, the order of the columns can't be changed**

Reviewer: 👍

- **reduced_penetrance(str)** If a gene is known to have reduced penetrance this is annotated. *Optional*
- **database_entry_version(str)** The database entry version is a way to track when a a gene was added or modified. *Optional*

> Each gene in a gene panel have to be identified with a [HGNC](https://www.genenames.org/) id.

Reviewer: 👍

- **genetic_disease_models** Any of the following values: [AR,AD,XR,XD,MT,X,Y]
- **mosaicism** Any entry here will be interpreted as true
- **reduced_penetrance** Any entry here will be interpreted as true
- **database_entry_version** This should refer to a earlier version of the panel

## Uploading a new gene panel version

You can upload a text file in Scout to update an existing gene panel. It should follow the format specified in this file: [panel-example](../static/scout-3-panel-file-example.csv). This file is `;` (semicolon) separated. Tab is also an accepted file separator.

You can also use this example [Excel template](../static/scout-3-panel-file-example.xlsx) as a starting point. When you are ready to update the gene panel simply:

1. choose **"Save as..."** in Excel and select **"Comma Separated Values (.csv)"** as the format
2. fill out the form in Scout and upload the "*.csv" file to update your gene panel

> **When creating a new panel it is very important that id doesn't contain blank lines.**

Reviewer: ```suggestion
**When creating a new panel it is very important that it does not contain blank lines.**
```
